[DeviceMesh] Fix a corner case for coalesce in cute layout and mesh slicing

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.13.0) (oldest at bottom):
* __->__ #169454

While investigating https://github.com/pytorch/pytorch/issues/169381 I found two corner cases, one for coalesce and another one is for slicing.

1. When all sizes of a layout are 1, the coalesce will skip all of strides and return a layout of (1, 0) which is wrong. So we don't start with 0 as the stride but with the last stride. Modifying unit test to test all expected coalesced layout so that we can verify the correctness of the code. Previous unit test didn't capture this case.
2. When the size of mesh is smaller than world size, some ranks could have empty dim_group_names this will cause errors when slicing, to support that, we add a skip. Also adding a unit test for it.

fduwjj committed:
[DeviceMesh] Fix a corner case for coalesce in cute layout and mesh slicing

[ghstack-poisoned]

lw commented on torch/distributed/_pycute/layout.py:
```diff
@@ -167,12 +167,13 @@ def coalesce(layout: Layout, profile: LayoutProfile = None) -> Layout:
+    flattened_stride = flatten(layout.stride)
     result_shape = [1]
-    result_stride = [0]
+    result_stride = [flattened_stride[-1]]
```
You said

> When all sizes of a layout are 1, the coalesce will skip all of strides and return a layout of (1, 0) which is wrong

I disagree with that claim. One of the "purposes" of coalescing is also to normalize a layout and produce a canonical representation. This is key because we (might) use the layout as a cache key hence we need to be able to deduplicate layouts which are effectively equivalent.

If we use the last stride instead of 0, we might break this property. For example, layouts (1, 1):(42, 3) and (1, 1, 1):(17, 4, 13) are, in fact, the same layout, but after this PR they coalesce to two _different_ layouts (i.e., (1,):(3,) and (1,):(13,) respectively). This is not acceptable.

Note, also, that if the size of a layout is (1,), then the stride is _irrelevant_, and any value is acceptable. Which I believe is why the original code was using 0 as a "default" stride.

Now, I wonder why this "fix" was needed in the first place. Does another part of the code fail when given a stride of 0? If so, we likely need to fix that piece of code instead...

lw replied:
I'll take a look at the original issue tomorrow, but I suspect it can be fixed more robustly by coalescing `slices_sizes` and `slices_strides` here: https://github.com/pytorch/pytorch/blob/7348cb355ff0a6f79cd4871215aea72185748734/torch/distributed/device_mesh.py#L831-L832

Since the error you encountered would only be raised if a 0-stride dimension occurs in the middle of a layout, but this is precisely what coalesce helps prevent.

Note also that the whole check could also just be removed (there's a TODO which suggests doing just that), we simply haven't done it yet because we don't want to expand the API surface without a use-case.

wconstab replied:
@fduwjj did you add this test case? i think it sounds like a good idea

> If here we take stride[-1] instead of 0, we might break this property. For example, layouts (1, 1):(42, 3) and (1, 1, 1):(17, 4, 13) are, in fact, the same layout, but after this PR they coalesce to two different layouts (i.e., (1,):(3,) and (1,):(13,) respectively). This is not acceptable. (In fact, we might want to add a test case to check this property)

Also, if the only problem that is caused by a 0 stride was tripping this assertion, can't we just workaround to special case stride=0 when implementing the assertion?

lw reviewed (changes requested):
I'm a bit suspicious of at least one of the fixes and would like to understand more before approving.

fduwjj commented on torch/distributed/device_mesh.py:
```diff
@@ -675,21 +675,17 @@ def _create_sub_mesh(
-                    )
+                    idx = not_none(self._mesh_dim_names).index(name)
+                    if idx < len(self._dim_group_names):
+                        slice_dim_group_name.append(self._dim_group_names[idx])
```
Well, we will still need this inside else for flattened mesh names as well...

fduwjj committed:
Update on "[DeviceMesh] Fix a corner case for coalesce in cute layout and mesh slicing"


While investigating https://github.com/pytorch/pytorch/issues/169381 I found two corner cases, one for coalesce and another one is for slicing.

1. When all sizes of a layout are 1, the coalesce will skip all of strides and return a layout of (1, 0) which is wrong. So we don't start with 0 as the stride but with the last stride. Modifying unit test to test all expected coalesced layout so that we can verify the correctness of the code. Previous unit test didn't capture this case.
2. When the size of mesh is smaller than world size, some ranks could have empty dim_group_names this will cause errors when slicing, to support that, we add a skip. Also adding a unit test for it.


[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Fix a corner case for coalesce in cute layout and mesh slicing"


While investigating https://github.com/pytorch/pytorch/issues/169381 I found two corner cases, one for coalesce and another one is for slicing.

1. When all sizes of a layout are 1, the coalesce will skip all of strides and return a layout of (1, 0) which is wrong. So we don't start with 0 as the stride but with the last stride. Modifying unit test to test all expected coalesced layout so that we can verify the correctness of the code. Previous unit test didn't capture this case.
2. When the size of mesh is smaller than world size, some ranks could have empty dim_group_names this will cause errors when slicing, to support that, we add a skip. Also adding a unit test for it.


[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Fix a corner case for coalesce in cute layout and mesh slicing"


While investigating https://github.com/pytorch/pytorch/issues/169381 I found two corner cases, one for coalesce and another one is for slicing.

1. When all sizes of a layout are 1, the coalesce will skip all of strides and return a layout of (1, 0) which is wrong. So we don't start with 0 as the stride but with the last stride. Modifying unit test to test all expected coalesced layout so that we can verify the correctness of the code. Previous unit test didn't capture this case.
2. When the size of mesh is smaller than world size, some ranks could have empty dim_group_names this will cause errors when slicing, to support that, we add a skip. Also adding a unit test for it.


[ghstack-poisoned]

fduwjj committed:
Update on "[DeviceMesh] Fix a corner case for coalesce in cute layout and mesh slicing"


While investigating https://github.com/pytorch/pytorch/issues/169381 I found two corner cases, one for coalesce and another one is for slicing.

1. When all sizes of a layout are 1, the coalesce will skip all of strides and return a layout of (1, 0) which is wrong. So we don't start with 0 as the stride but with the last stride. Modifying unit test to test all expected coalesced layout so that we can verify the correctness of the code. Previous unit test didn't capture this case.
2. When the size of mesh is smaller than world size, some ranks could have empty dim_group_names this will cause errors when slicing, to support that, we add a skip. Also adding a unit test for it.


[ghstack-poisoned]

fduwjj commented on test/distributed/_pycute/test_coalesce.py:
```diff
@@ -47,65 +47,82 @@
+
+        layout = Layout((1, 1, 1), (17, 4, 3))
+        expected_coalesced_layout = Layout(1, 0)
+        self.helper_test_coalesce(layout, expected_coalesced_layout)
```
@wconstab the tests @lw asked is added here.

fduwjj committed:
Update on "[DeviceMesh] Fix a corner case for coalesce in cute layout and mesh slicing"


While investigating https://github.com/pytorch/pytorch/issues/169381 I found two corner cases, one for coalesce and another one is for slicing.

1. When all sizes of a layout are 1, the coalesce will skip all of strides and return a layout of (1, 0) which is wrong. So we don't start with 0 as the stride but with the last stride. Modifying unit test to test all expected coalesced layout so that we can verify the correctness of the code. Previous unit test didn't capture this case.
2. When the size of mesh is smaller than world size, some ranks could have empty dim_group_names this will cause errors when slicing, to support that, we add a skip. Also adding a unit test for it.


[ghstack-poisoned]

lw commented on torch/distributed/device_mesh.py:
```diff
@@ -676,23 +676,30 @@ def _create_sub_mesh(
-                        ]
-                    )
+            if len(self._dim_group_names) > 0:
+                assert len(self._dim_group_names) == len(
```
Would it cause any problems to move this assert just before the `if`? This way it would also cover the case of both lists being empty.

fduwjj commented on torch/distributed/_mesh_layout.py:
```diff
@@ -104,7 +104,15 @@ def coalesce(self) -> "_MeshLayout":
+            non_zero_strides = tuple(s if s != 0 else 1 for s in layout.stride)
+            return _MeshLayout(layout.shape, non_zero_strides)
+        else:
+            return _MeshLayout(layout.shape, layout.stride)
```
I think it is not just about `pruning all size-1 dims from the layout`. It is also about when we do flatten among certain dims and the rest of dims are still there. So you can not remove all these size-1 dims because they are actual dims of the mesh.

lw reviewed (approved):
I'm ok to go with stride=1 instead of stride=0 as a conventional value for size=1 dims. It seems harmless enough and preserves the "normalization" property of coalescing.

fduwjj committed:
Update

[ghstack-poisoned]
